vector/src/main/java/im/vector/adapters/AutoCompletedUserAdapter.java

```diff
@@ -0,0 +1,187 @@
+     * @param session          the session
+     * @param users            the users list
+     */
+    public AutoCompletedUserAdapter(Context context, int layoutResourceId, MXSession session, Collection<User> users) {
+        super(context, layoutResourceId);
+        mContext = context;
+        mLayoutResourceId = layoutResourceId;
+        mLayoutInflater = LayoutInflater.from(mContext);
+        mSession = session;
+        addAll(users);
+        mUsersList = new ArrayList<>(users);
+    }
+
+    @Override
+    public View getView(int position, View convertView, ViewGroup parent) {
+        if (convertView == null) {
+            convertView = mLayoutInflater.inflate(mLayoutResourceId, parent, false);
+        }
+
+        User user = getItem(position);
+
+        VectorCircularImageView avatarView = (VectorCircularImageView) convertView.findViewById(R.id.item_user_auto_complete_avatar);
+        TextView userNameTextView = (TextView) convertView.findViewById(R.id.item_user_auto_complete_name);
+
+        VectorUtils.loadUserAvatar(mContext, mSession, avatarView, user.getAvatarUrl(), user.user_id, user.displayname);
+
+        if (!mIsSearchingMatrixId) {
+            userNameTextView.setText(user.displayname);
+        } else {
+            userNameTextView.setText(user.user_id);
+        }
+
+        return convertView;
+    }
+
+    @Override
+    public android.widget.Filter getFilter() {
+        if (mFilter == null) {
+            mFilter = new AutoCompletedUserFilter();
+        }
+        return mFilter;
+    }
+
+    private class AutoCompletedUserFilter extends android.widget.Filter {
+        @Override
+        protected FilterResults performFiltering(CharSequence prefix) {
+            FilterResults results = new FilterResults();
+            List<User> newValues;
+
+            if (prefix == null || prefix.length() == 0) {
+                newValues = new ArrayList<>();
+                mIsSearchingMatrixId = true;
+            } else {
+                newValues = new ArrayList<>();
+                String prefixString = prefix.toString().toLowerCase();
+                mIsSearchingMatrixId = prefixString.startsWith("@");
+
+                if (mIsSearchingMatrixId) {
+                    for (User user : mUsersList) {
+                        if ((null != user.user_id) && user.user_id.toLowerCase().startsWith(prefixString)) {
+                            newValues.add(user);
+                        }
+                    }
+                } else {
+                    for (User user : mUsersList) {
+                        if ((null != user.displayname) && user.displayname.toLowerCase().startsWith(prefixString)) {
+                            newValues.add(user);
+                        }
+                    }
+                }
+            }
+
+            // sort the results
+            if (mIsSearchingMatrixId) {
+                Collections.sort(newValues, new Comparator<User>() {
```

Comment:
Comparator could be created only once outside the filter class instead of being recreated at each filtering